### PR TITLE
Fix internal assert custom function when input does not require grad

### LINKED ISSUE #71119: Custom function forward-mode AD internal assert when returning same tensor
```python
import torch.autograd.forward_ad as fwAD
import torch

class Test(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x

    @staticmethod
    def jvp(ctx, x_t):
        return x_t

x = torch.tensor(1.)
t = torch.tensor(1.)

with fwAD.dual_level():
    dual_x = fwAD.make_dual(x, t)
    Test.apply(dual_x)
```

cc @ezyang @gchanan @zou3519 @bdhirsh @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72133
* __->__ #72008

Fixes  #71119

Technically BC-breaking because when an input does not require grad, previously it was returned as-is instead of a view because it didn't need to. Now we will also return a view in that case (whether or not forward AD runs).

Differential Revision: [D33859553](https://our.internmc.facebook.com/intern/diff/D33859553)

### CODE REVIEW

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

File: torch/csrc/autograd/custom_function.cpp [hunk 1]
```diff
@@ -215,6 +215,13 @@ optional_variable_list _process_backward_mode_ad(
                          bool is_differentiable) {
     if (!is_differentiable) {
       if (!var.requires_grad()) {
+        if (is_input) {
+          // We need this in case forward AD runs, since forward AD assumes that the output
+          // processed by backward ad can never be an input as-is.
+          // If forward AD does not run, we do some extra work here, but that is OK
+          AutoGradMode grad_mode(false);
```
Comment (albanD) [inline torch/csrc/autograd/custom_function.cpp:222]: - Can we have this in a separate function that is called here and below?
- Didn't we add a context to disable forward mode AD while doing this view in a previous PR??

Comment (soulitzer) [reply torch/csrc/autograd/custom_function.cpp:222]: - sounds good
- that was the initial plan in the previous PR, but we eventually decided not to in favor of just returning early in the _process_forward_ad function. (we did disable forward AD while calling user-defined jvp)

Comment (soulitzer) [reply torch/csrc/autograd/custom_function.cpp:222]: Hmm maybe I am missing something? My understanding is that we had two options:
1. if forward is no-op, we require user to manually return a view in jvp (otherwise error) to match the behavior of what backward does (when the output does require grad)
2. if forward is no-op, automatically return a view in jvp (we did this by continuing to run the view with forward grad enabled, and just adding a `continue` as we loop through the outputs in _process_forward_ad, as to avoid setting the forward grad again).

We ended up going with (2), but we didn't consider the case in _process_backward_ad when the output is considered non-differentiable (perhaps because it does not require grad), so in this PR, we add performing the view in this case as well.

File: test/test_autograd.py [hunk 2]
```diff
@@ -5588,7 +5588,9 @@ def jvp(ctx, nt, x_t, nt2, y_t):
                   check_backward_ad=False, check_batched_grad=False)
 
     def test_custom_function_forward_mode_forward_is_no_op(self):
-        for jvp_mul_by_2 in (True, False):
+        error_regex = "A custom Function's forward is returning a view \\(or an input as-is\\)"
+
+        for returns_view in (True, False):
```
Comment (albanD) [inline test/test_autograd.py:5593]: Why not add a 3rd case instead of removing one of the existing ones?

**Review by albanD — APPROVED** [review]
LGTM!

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33859553).

**Review by albanD — APPROVED** [review]
Thanks for the update!

**Comment by anjali411** [comment]
@soulitzer could you please add the release notes labels?
